rfc/plugins.md

```diff
@@ -0,0 +1,248 @@
+
+Tightly-coupled service and plugin releases is a poor developer and operator story, but an acceptable shortcoming for an initial release. By using Front50 as the source of truth, we’ll be fairly well equipped to make an acceptable runtime plugin loading story.
+
+Assuming a world where in-process plugins can be upgraded at runtime, services will need a background process to download these new plugins whenever a delta is detected in Front50, and the plugin refreshed. PF4J offers an abstraction for this reloading mechanism, but we would need to figure out how Spring integrations are handled. For example, it’s possible we may need to make the constraint that exported beans from a plugin are not refreshed on plugin up/down grade at least initially.
+
+#### Remote
+Remote plugin upgrade process is not much different than iterating a remote service, since that’s what it is.
+
+When a remote plugin is released, its internal functionality will change immediately, just as it would be expected for a remote service. When the plugin starts, it must send a registration event to Spinnaker. This registration event would contain, at least:
+
+* Plugin ID
+* Other plugin metadata (author, description, version, etc).
+* What extension points it integrates with
+
+Upon receiving this registration event, Spinnaker would then update the Front50 registry with the new information and downstream services would re-bootstrap the plugin, adding or removing it from extension points as necessary.
+
+### Dependencies
+A series of new kork modules will be provided for libraries and services alike to consume to expose new extension points. Under the covers, this introduces PF4J as a new global service dependency.
+
+In-process plugins, by default, will have their own ClassLoader that is isolated from other Plugin ClassLoaders. These Plugin ClassLoaders will be children of the root ClassLoader, meaning they cannot have dependencies that conflict with the root ClassLoader. Future work may be performed to allow for different levels of isolation if necessary.
+
+**TBD**: We would like to offer different plugin runtimes besides in-process as plugins themselves. These plugin runtimes would then come with their own dependencies that would be exposed to the parent service, such as gRPC or Lambda function invocation.
+
+## Drawbacks
+There are a handful major issues with the proposed direction:
+
+1. In-process plugins have a poor developer and operator experience which can negatively affect end-users, especially in rollback scenarios. In-process plugins also require developers to have more knowledge of the Spinnaker toolchain and potentially the service architecture. Exposing logging, telemetry and crash reports to plugin developers for in-process plugins will be difficult or impossible dependent on a Spinnaker installation’s preferences.
```

Comment:
Yes, but that would first require a bugsnag plugin. Where does bugsnag send its data if the plugin isn't working correctly?

Reply:
if which plugin isn't working? the bugsnag client not working? That's like saying what if elk/sumo is down, no? BugSnag (and similar services) have a ton resilience and store data locally until it can be sent, etc. but yes, the spinnaker plugin developer would have to load bugsnag/or equivalent and if the plugin fails early on it might not have initialized bugsnag yet...

Reply:
After thinking on this more, I like the essence of what you're getting at. Perhaps there's room for a diagnostics plugin of sorts, if you want to call it that, that can be implemented to report telemetry, logs and errors to a separate system. The tricky bit here is each plugin developer is going to want to consume this information in a way that suits them; which would mean varying diagnostics plugins floating around depending on the various plugins that are enabled.

Am I going off the rails here? Is there merit in taking this thought further?

Reply:
FWIW, I think the plugin developer vs. the Spinnaker operator needs are very different here. What you guys discuss above makes sense for developers. For operators, I would expect all plugins to tap into the existing logging/monitoring infrastructure used by Spinnaker core components, in other words, it shouldn't any burden just because you added plugins to the mix.

Reply:
Yes, absolutely, in-process plugins will be plugged into the existing telemetry / logging system, but that's not useful for the plugin developers: Take for example Netflix: No one outside of Delivery Engineering has access to the Spinnaker logs. We don't want to be in the business of copy/pasting logs to plugin developers whenever something goes sideways for something they developed.